substrate/runtime/primitives/src/traits.rs

```diff
@@ -147,10 +149,86 @@ impl<A: Executable, B: Executable> Executable for (A, B) {
 	}
 }
 
+/// Abstraction around hashing
+pub trait Hashing {
+	/// The hash type produced.
+	type Output;
+
+	/// Produce the hash of some byte-slice.
+	fn hash(s: &[u8]) -> Self::Output;
+
+	/// Produce the hash of some codec-encodable value.
+	fn hash_of<S: codec::Slicable>(s: &S) -> Self::Output {
+		codec::Slicable::using_encoded(s, Self::hash)
+	}
+
+	/// Produce the patricia-trie root of a mapping from indices to byte slices.
+	fn enumerated_trie_root(items: &[&[u8]]) -> Self::Output;
+
+	/// Iterator-based version of `enumerated_trie_root`.
+	fn ordered_trie_root<
+		I: IntoIterator<Item = A>,
+		A: AsRef<[u8]>
+	>(input: I) -> Self::Output;
+
+	/// The Patricia tree root of the given mapping as an iterator.
+	fn trie_root<
+		I: IntoIterator<Item = (A, B)>,
+		A: AsRef<[u8]> + Ord,
+		B: AsRef<[u8]>
+	>(input: I) -> Self::Output;
+
+	/// Acquire the global storage root.
+	fn storage_root() -> Self::Output;
```

Comment:
that doesn't seem like something that belongs in a hash trait, rather it should be some functionality which is parameterized by it

Reply:
was already there and being used, so if there's any refactoring to be done, i would prefer to leave it for another pr.